File: monitoring/perf_level.cc
//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
//  This source code is licensed under both the GPLv2 (found in the
//  COPYING file in the root directory) and Apache 2.0 License
//  (found in the LICENSE.Apache file in the root directory).
//

#include <assert.h>

#include "monitoring/perf_level_imp.h"

namespace ROCKSDB_NAMESPACE {

thread_local PerfLevel perf_level = kEnableCount;

void SetPerfLevel(PerfLevel level) {
  assert(level > kUninitialized);
  assert(level < kOutOfBounds);
  perf_level = level;
}

PerfLevel GetPerfLevel() { return perf_level; }

}  // namespace ROCKSDB_NAMESPACE

File: java/src/main/java/org/rocksdb/PerfLevel.java
//  Copyright (c) Meta Platforms, Inc. and affiliates.
//  This source code is licensed under both the GPLv2 (found in the
//  COPYING file in the root directory) and Apache 2.0 License
//  (found in the LICENSE.Apache file in the root directory).

package org.rocksdb;

public enum PerfLevel {
  /**
   * Unknown setting
   */
  UNINITIALIZED((byte) 0),
  /**
   * disable perf stats
   */
  DISABLE((byte) 1),
  /**
   * enable only count stats
   */
  ENABLE_COUNT((byte) 2),
  /**
   * Other than count stats, also enable wait/delay time stats for
   * user threads blocked in RocksDB
   */
  ENABLE_WAIT((byte) 3),
  /**
   * Also enable time stats for write operations (WAL write time,
   * scheduling time, etc.) without the overhead of read-path timers
   */
  ENABLE_TIME_FOR_WRITE((byte) 4),
  /**
   * Also enable time stats for all operations except for mutexes
   */
  ENABLE_TIME_EXCEPT_FOR_MUTEX((byte) 5),
  /**
   * Other than time, also measure CPU time counters. Still don't measure
   * time (neither wall time nor CPU time) for mutexes
   */
  ENABLE_TIME_AND_CPU_TIME_EXCEPT_FOR_MUTEX((byte) 6),
  /**
   * enable count and time stats
   */
  ENABLE_TIME((byte) 7),

  /**
   * Do not use
   * @deprecated It's here to just keep parity with C++ API.
   */
  @Deprecated OUT_OF_BOUNDS((byte) 8);

  PerfLevel(byte _value) {
    this._value = _value;
  }

  private final byte _value;

  public byte getValue() {
    return _value;
  }

  public static PerfLevel getPerfLevel(byte level) {
    for (PerfLevel l : PerfLevel.values()) {
      if (l.getValue() == level) {
        return l;
      }
    }
    throw new IllegalArgumentException("Uknknown PerfLevel constant : " + level);
  }
}
